lib/quicrq.c

```diff
@@ -1788,6 +1789,101 @@ int quicrq_receive_stream_data(quicrq_stream_ctx_t* stream_ctx, uint8_t* bytes,
+/*
+ * todo: add documentation
+ */
+int quicrq_receive_warp_or_rush_stream_data(quicrq_cnx_ctx_t* cnx_ctx, quicrq_uni_stream_ctx_t* uni_stream_ctx, uint8_t* bytes, size_t length, int is_fin) {
+    // warp-header:grp-id [obj1 , obj2 ]
+    int ret = 0;
+
+    while (ret == 0 && length > 0) {
+        /* There may be a set of messages back to back, and all have to be received. */
+        if (uni_stream_ctx->receive_state == quicrq_receive_done) {
+            /* Protocol violation, was not expecting any message */
+            ret = -1;
+            break;
+        }
+        else {
+            /* Receive the next message on the stream, if any */
+            int is_finished = 0;
+            uint8_t* next_bytes = quicrq_msg_buffer_store(bytes, length, &uni_stream_ctx->message_buffer, &is_finished);
+            if (next_bytes == NULL) {
+                /* Something went wrong */
+                ret = -1;
+            }
+            else
+            {
+                length = (bytes + length) - next_bytes;
+                bytes = next_bytes;
+                if (is_finished) {
+                    /* Decode the incoming message */
+                    quicrq_message_t incoming = { 0 };
+                    const uint8_t* r_bytes = quicrq_msg_decode(uni_stream_ctx->message_buffer.buffer, uni_stream_ctx->message_buffer.buffer + stream_ctx->message_receive.message_size, &incoming);
+
+                    if (r_bytes == NULL) {
+                        /* Message was incorrect */
+                        ret = -1;
+                    } else {
+                        switch (incoming.message_type) {
+                            case QUICRQ_ACTION_WARP_HEADER:
+                                uni_stream_ctx->current_group_id = incoming.group_id;
+                                uni_stream_ctx->receive_state = quicrq_receive_warp_header;
+                                quicrq_stream_ctx_t* ctrl_stream_ctx = quicrq_get_control_stream_for_media_id(cnx_ctx, incoming.media_id);
+                                uni_stream_ctx->receive_state = quicrq_receive_warp_header;
+                                uni_stream_ctx->control_stream_ctx = ctrl_stream_ctx;
+                                /* todo: add check for null control stream */
+                                quicrq_log_message(cnx_ctx, "UniStream %" PRIu64 ", received warp header message uni_id= %" PRIu64,
+                                                   ", ControlStream id= %" PRIu64 " media id=%" PRIu64,
+                                                   uni_stream_ctx->stream_id, uni_stream_ctx->control_stream_ctx->stream_id, incoming.media_id);
+
+                                break;
+                            case QUICRQ_ACTION_OBJECT_HEADER:
+                                if (uni_stream_ctx->receive_state != quicrq_receive_warp_header) {
+                                    /* Protocol error */
+                                    ret = -1;
+                                }
```

Comment:
Maybe review the name of your states. "quicrq_receive_warp_header" kind of imply that you are waiting for `QUICRQ_ACTION_WARP_HEADER`, but in fact you use it when waiting for WARP object. Also, the states should be:

1- get notified that stream is open
2- receive warp header
3- receive object headers:
      - receive object or FIN mark
      - if object received, process it
      - if FIN received, close the stream.
      - if final group and object known, and if everything was received up to final_group, final_object, mark the control stream active and schedule sending FIN.

Reply:
Can we work on this live ? I get stuck in the right API calls on some of these ..